### DISCUSSION TITLE
Wiggler calculation with low number of periods

### DISCUSSION THREAD

**Discussion by swhite2401** [atcollab/at#871] [Python AT]
@lfarv @oscarxblanco @ZeusMarti 

When setting the number of periods below 15 over a length of 1m there seem to be a problem with the calculation since optics is strongly distorted in both planes.

Do you know of any limitation in this respect? Did I miss something?

Thanks!

**Comment by ZeusMarti** [comment]
Hi @swhite2401,

I did the following quick test:


```
B=3; % Field in Tesla
Energy=3e9; % Energy in eV
L_wiggler=1; % length in m
nper_max=30;
nper_min=2;
npers=nper_min:nper_max; % Number of periods to be tested
Rin=1e-4*ones(6,1); % input particle at 100 mum, 100 murad and 0.01%
Tout=zeros(6,numel(npers));

for ii=1:numel(npers)
    ID = atwiggler('WiggHarm', L_wiggler, L_wiggler/npers(ii), B, Energy,...
        'GWigSymplecticRadPass','Nstep',71);
    Tout(:,ii)=linepass({ID},Rin);
end

fact=[1e3 1e3 1e3 1e3 1e2 1e3];
labls={'x [mm]','p_x [mrad]','y [mm]','p_y [mrad]','\delta [%]','ct [mm]'};
for ii=1:6
    subplot(3,2,ii);
    plot(npers,fact(ii)*Tout(ii,:),'-r.');
    xlabel('Periods no.');
    ylabel(labls{ii})
end

```
The result is the following with `GWigSymplecticPass`:
![Track_1mWiggler](https://github.com/user-attachments/assets/3d4128a5-3bae-45f9-a05a-d9bf09bc1ff8)
And with `GWigSymplecticRadPass`:
![Track_1mWiggler_Rad](https://github.com/user-attachments/assets/fc0c6555-e2bf-47f5-9054-9d28085e54df)

While there is some suspicious behavior in the case of `GWigSymplecticRadPass`, that we should certainly study in further detail, I see no significantly different behavior around 15 periods. Just to be sure: Are you sure the difference you see comes only from the wiggler element? Is the ring length constant when adding the wiggler? 

Cheers,
Zeus

**Reply by swhite2401** [reply]
Thanks for the reply! Let me try to put together a short example to show you what the problem is. But to summarize what I am doing is simply change the number of periods of the wiggler, nothing else changes, and below 15 there seem to be some kind of step in the behavior.
I am looking at optics (betax, betay), maybe there is some issue in the optics calculation in the presence of the wiggler?

**Comment by swhite2401** [comment]
Hello @ZeusMarti ,

After some more investigation, it seems that the wiggler element is fine but not compatible with the `plot_beta()` function, probably due to the slicing. Below the scripts I used for testing and the figures it generates. Using get_optics() results are the same but when using plot_beta() the differ significantly.
@lfarv, shall we adapt the plot function?



```
import at
import matplotlib.pyplot as plt

ring = at.load_lattice('./hmba.mat', periodicity=32)

l_wiggler = 1.0
nperiods = 16
b_wiggler = 1.5
wiggler = at.Wiggler('DW', l_wiggler, l_wiggler/nperiods, b_wiggler, energy=ring.energy)


ring = ring.rotate(ring.get_uint32_index('CellCenter')[0])

dr01_idx  = ring.get_uint32_index('DR_01')[0]
dr01 = ring[dr01_idx]
dr01_wig = dr01.insert((((dr01.Length-wiggler.Length/2.0)/dr01.Length, wiggler), ))
ring = ring[:dr01_idx] + dr01_wig + ring[dr01_idx+1:]

dr45_idx  = ring.get_uint32_index('DR_45')[0]
dr45 = ring[dr45_idx]
dr45_wig = dr45.insert(((wiggler.Length/2/dr45.Length, wiggler), ))
ring = ring[:dr45_idx] + dr45_wig + ring[dr45_idx+1:]

wigglers = ring.get_elements(at.Wiggler)


for e in wigglers:
    e.Lw = l_wiggler/16  
_, _, ld16 = ring.get_optics(refpts=at.All)
ring.plot_beta()

for e in wigglers:
    e.Lw = l_wiggler/10 
_, _, ld10 = ring.get_optics(refpts=at.All)
ring.plot_beta()

plt.figure()
plt.plot(ld16.beta)
plt.plot(ld10.beta, linestyle='dashed')

plt.show()
```

![plot_beta_10](https://github.com/user-attachments/assets/e12fbc86-36b6-467d-a571-6c0e10c3ef78)
![plot_beta_16](https://github.com/user-attachments/assets/ce734dee-2b5e-457d-a12b-0d71eda4f11a)
![get_optics](https://github.com/user-attachments/assets/4dc22a63-fb9b-4bbe-a9d3-b47e5834c71c)
